Fixes for bumpfee and minimum fees (use incrementalRelayFee for BIP 125 (RBF) replacement)

morcos opened the pull request:
This commit includes several minor fixups from #9380.

Most importantly the change to `AcceptToMemoryPoolWorker` to use `incrementalRelayFee`was an oversight that was left out of #9380.  

`bumpfee` is adapted for that change and the changes from #9380,
Also includes fixes for #8456 

Please tag as a bug fix for 0.14

Edited to reflect additional changes.

morcos committed:
Use incrementalRelayFee for BIP 125 replacement

morcos committed:
Fix missing use of dustRelayFee

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2652,8 +2652,8 @@ UniValue bumpfee(const JSONRPCRequest& request)
             "The user can specify a confirmation target for estimatefee.\n"
             "Alternatively, the user can specify totalFee, or use RPC setpaytxfee to set a higher fee rate.\n"
-            "At a minimum, the new fee rate must be high enough to pay a new relay fee (relay fee amount returned\n"
-            "by getnetworkinfo RPC) and to enter the node's mempool.\n"
```
Would it be a good idea to update getnetworkinfo to return the incremental relay fee?

morcos replied:
I think from a user perspective, it might be easier to keep incremental relay fee as something that is kind of unknown to them and is effectively just the hard coded 1000 sat / KB, however if we ever think it makes sense to have people changing it (or we think the default needs changing it), then it would be time to expose it.  I realize that's not a perfect solution, but I also think confusing people with too many random minimums isn't ideal either...    At least the error message for bumpfee outputs what the needed increment is..

ryanofsky replied:
I think it'd be good to have this if only so this documentation can be less vague. If I'm trying to write an RPC call that passes a totalFee and the documentation just says I need to "pay a new relay fee", what am I supposed to do at the point to find out what value I need to pass for the call to complete reliably? Search for "bitcoin relay fee"? Look at the source code? I think I'd be better off with documentation that says I need to increase the fee by at least the amount returned in getnetworkinfo.

morcos replied:
Ok done

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2834,7 +2834,7 @@ UniValue bumpfee(const JSONRPCRequest& request)
     // If the output would become dust, discard it (converting the dust to fee)
     poutput->nValue -= nDelta;
-    if (poutput->nValue <= poutput->GetDustThreshold(::minRelayTxFee)) {
+    if (poutput->nValue <= poutput->GetDustThreshold(dustRelayFee)) {
```
Not directly related, but CTxOut IsDust and GetDustThreshold methods still take arguments named minRelayTxFee. Would seem like a good idea to rename those to dustRelayFee.

morcos replied:
yes, that has annoyed me a few times as well and I was going to change it, but @jtimon has #9279 which I was hoping would eventually get merged and make it go away all together.

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2804,9 +2804,9 @@ UniValue bumpfee(const JSONRPCRequest& request)
         }
 
-        // new fee rate must be at least old rate + minimum relay rate
-        if (nNewFeeRate.GetFeePerK() < nOldFeeRate.GetFeePerK() + ::minRelayTxFee.GetFeePerK()) {
```
Consider keeping the :: prefix for global variables. As someone new to the code who doesn't have all the global variables memorized, I think the prefix is very helpful for readability, especially in a large function like bumpfee with similar sounding local variables.

morcos replied:
ok i don't mind doing that, but will just do for the changes introduced in this PR here?

ryanofsky replied:
Great, that sounds fine to me.

ryanofsky reviewed (approved):
utACK fa005cce2b7a8e02cedc7356d2d1933b94c26c39

MarcoFalke commented:
utACK fa005cc, agree with the nits.

morcos commented:
Only change is adding `::` to global variables

morcos committed:
Fix to have miner test aware of new separate block min tx fee

morcos committed:
[rpc] Add incremental relay fee to getnetworkinfo

morcos commented:
Took @ryanofsky's suggestion to add the incremental fee to getnetworkinfo and also added a commit that fixed an issue with block min tx fee in a test.

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2652,8 +2652,8 @@ UniValue bumpfee(const JSONRPCRequest& request)
-            "At a minimum, the new fee rate must be high enough to pay a new relay fee over the original fee\n"
-            "to enter the node's mempool.\n"
+            "At a minimum, the new fee rate must be high enough to pay an additional new relay fee (incrementalfee\n"
+            "returned by getnetowrkinfo) to enter the node's mempool.\n"
```
typo getnetowrkinfo

ryanofsky reviewed (approved):
utACK 72af459a53f83d507aa83c93cd3ef8aad2e1c7e0

morcos commented:
typo fixed

ryanofsky commented:
re-utACK fe8e8efcf91fa92db68aabeb0a1709b032e60dd6

TheBlueMatt commented:
utACK fe8e8efcf91fa92db68aabeb0a1709b032e60dd6

jtimon commented:
per-commit utACK: https://github.com/bitcoin/bitcoin/pull/9589/commits/de6400de5d4b654007f21901d9555ddea0face82 https://github.com/bitcoin/bitcoin/pull/9589/commits/fe8e8efcf91fa92db68aabeb0a1709b032e60dd6 https://github.com/bitcoin/bitcoin/pull/9589/commits/6b331e6cf977caa20c83a28a8c6c2af0b6c451c5

MarcoFalke commented:
utACK fe8e8efcf91fa92db68aabeb0a1709b032e60dd6

sdaftuar commented:
Perhaps also update the comments in `CTransaction::GetDustThreshold()`?  The comment there is pretty old I guess, referring to CTransaction::minRelayTxFee.  Anyway some explanation of the dustRelayFee would be helpful there.

sdaftuar commented:
This is minor, but this comment could be fixed in `policyestimator_tests.cpp`, line 190 (`minRelayTxFee` should be `incrementalRelayFee`):
```    
// evict that transaction which should set a mempool min fee of minRelayTxFee + feeV[5]
```

sdaftuar commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2804,9 +2804,9 @@ UniValue bumpfee(const JSONRPCRequest& request)
-        if (nNewFeeRate.GetFeePerK() < nOldFeeRate.GetFeePerK() + ::minRelayTxFee.GetFeePerK()) {
-            nNewFeeRate = CFeeRate(nOldFeeRate.GetFeePerK() + ::minRelayTxFee.GetFeePerK());
+        // new fee rate must be at least old rate + minimum incremental relay rate
+        if (nNewFeeRate.GetFeePerK() < nOldFeeRate.GetFeePerK() + ::incrementalRelayFee.GetFeePerK()) {
```
This change (while correct) exposes a bug in the logic, where we weren't otherwise comparing nNewFeeRate to either minTxFee or minRelayTxFee -- we should probably either add a call to compare against GetRequiredFee(), or rewrite the above to just use CWallet::GetMinimumFee.

Separately -- I now realize that we're also not comparing nNewFeeRate to maxTxFee either?  That seems like a separate bug in bumpfee...

morcos committed:
Use CWallet::GetMinimumFee in bumpfee

Use the wallet's fee calculation logic to properly clamp fee against minimums and maximums when calculating the fee for a bumpfee transaction.  Unless totalFee is explictly given, in which case, manually check against min, but do nothing to adjust given fee.

In all cases do a final check against maxTxFee (after adding any incremental amount).

morcos committed:
Refactor GetMinimumFee to give option of providing targetFee

morcos commented:
I added two new commits at the end to address the bugs @sdaftuar identified, hopefully this will help the prior review of the other commits remain useful.

I didn't bother addressing the nits, b/c I'd like to get this merged ASAP and want to concentrate on the important parts.

MarcoFalke commented:
re-ACK 9412ca4e7baba5130baaf173b5f189c240f1408a

sdaftuar commented:
I don't think we should ever allow maxTxFee to be violated.  Apart from the minor issue of slightly confusing semantics if maxTxFee is only enforced on totalFee and not an automatically generated one, the maxTxFee limit is the only thing preventing us from generating infinitely high (bounded only by the change output) fees if the user keeps calling bumpfee in a loop.  

(Moreover, I believe a fee in excess of maxTxFee will be rejected by the mempool, and currently bumpfee does not handle this gracefully -- I'll open another PR to clean up the handling in this case.)

morcos commented:
OK sorry for all the churn..   Modified last commit to check against maxTxFee in all cases

Diff from 9412ca4 is here: http://pastebin.com/C1db37i3

sdaftuar commented:
ACK e8021ec

MarcoFalke commented:
Concept ACK on the last diff. Thanks @sdaftuar for raising the issue with maxTxFee!

TheBlueMatt commented:
utACK e8021ec9193c7e8137f9716bcafd197a357a624e much nicer to use wallet's GetMinimumFee functions, thanks.

btcdrak reviewed (approved):
utACK e8021ec
